apps/s_client.c

```diff
@@ -2389,9 +2391,19 @@ int s_client_main(int argc, char **argv)
         }
         break;
     case PROTO_CONNECT:
-        if (!OSSL_HTTP_proxy_connect(sbio, host, port, proxyuser, proxypass,
-                                     0 /* no timeout */, bio_err, prog))
-            goto shut;
+	{
+	    /* Need to parse the actual connect target now */
+            char *thost, *tport;
+            /* connectstr was already validated, so we can assume success below */
+            BIO_parse_hostserv(connectstr, &thost, &tport,
```

Comment:
To avoid re-parsing why not just save the host and port that you already have above when parsing the proxystr?

Reply:
Fair enough - I was trying to make minimal changes to prove/fix the issue.
In the version below I strdup() the original values of port & host from connectstr, for use in the HTTP connect string sent to the proxy. They're freed when done.

Reply:
strdup isn't available on all platforms, so you should do ```OPENSSL_strdup``` and ```OPENSSL_free```.

Reply:
Apologies  - I did use `OPENSSL_strdup` & `OPENSSL_free` in the patch, but didn't write it up correctly above.

Reply:
I apologize for not looking at your code.  Nice work :)

Reply:
Each of the tree such strings (connectstr, proxystr, and bindstr) is parsed only once, so I don't see a re-parsing issue here.

Reply:
(I just found that my above comment was outdated.)